### PR TITLE
Check_enforcer enforce readme and examples check

### PR DESCRIPTION
**Pull request by crazygao**
# Description

Changes in readme.py: add a -c parameter to retrieve all path filters.
New utility fnmatch: same with fnmatch library, check if file name match some given glob.
Check enforcer: now when examples files are changed. It will run readme with -c parameter. get a valid list of path filter. check every file name against the path filter list and find out what pipeline should run.

This PR may need splitting.

refactor original PS1 script to Py script for GREAT PERF GAIN.

Take this PR as example: there are 58 files in the PR, and since it contains example changes, there are about 100 checks for each file.
Powershell will stuck on this 58*100 checks for over tens of minutes. At the same time, python could return me with correct results within seconds.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: scripts/building/check_enforcer.ps1 [hunk 1]
```diff
@@ -129,29 +121,78 @@ function sdk_cli_checks([ref]$failed_reason_ref, $valid_status_array) {
...
-            Write-Output "git diff --name-only HEAD main $_"
+function trigger_prepare($input_paths, [ref]$failed_reason_ref) {
+    foreach ($input_path in $input_paths) {
+        if ($checks.ContainsKey("samples_connections_connection")) {
+            continue
+        }
+        if ($input_path.Contains("examples") -or $input_path.Contains("samples")) {
+            # Define the input path
+            Push-Location $GithubWorkspace
+            $pipelines_samples = (python $GithubWorkspace/scripts/readme/readme.py -c | ConvertFrom-Json)
```
Comment (zhengfeiwang) [inline scripts/building/check_enforcer.ps1:132]: if one PR changes template, for example, remove path filters, then check enforcer will not guard any more?

File: scripts/building/check_enforcer.ps1 [hunk 2]
```diff
@@ -129,29 +121,78 @@ function sdk_cli_checks([ref]$failed_reason_ref, $valid_status_array) {
...
+        # input pattern /*: input_path should match last but one
+        # other input pattern: input_path should match last
+        $keys = $reverse_checks.Keys
+        $keys = $keys | Where-Object { (python $GithubWorkspace/scripts/building/fnmatch.py -g $_ -f $input_path) -eq "True" }
+        # $reverse_checks.Keys and $keys
+        foreach ($key_item in $keys) {
+            foreach ($key in $reverse_checks[$key_item]) {
+                if ($pipelines.ContainsKey($key)) {
+                    if ($special_care.ContainsKey($key)) {
+                        $pipelines[$key] = $special_care[$key]
```
Comment (zhengfeiwang) [inline scripts/building/check_enforcer.ps1:171]: intuitively, we need a preprocess, to make the key as path, and value is the triggered pipeline, so that this loop can be much smaller?

File: .github/workflows/samples_configuration.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ env:
   IS_IN_CI_PIPELINE: "true"
 
 jobs:
-  samples_notebook_ci:
+  samples_configuration:
```
Comment (zhengfeiwang) [inline .github/workflows/samples_configuration.yml:19]: maybe we can make this a separate PR to do so?

File: scripts/building/check_enforcer.ps1 (hunk 1 above)
Comment (crazygao) [reply scripts/building/check_enforcer.ps1:132]: Of course~ this is expected.

File: scripts/building/check_enforcer.ps1 (hunk 2 above)
Comment (crazygao) [reply scripts/building/check_enforcer.ps1]: Changed.

File: .github/workflows/samples_configuration.yml (hunk 3 above)
Comment (crazygao) [reply .github/workflows/samples_configuration.yml:19]: I use this part for testing, changing is OK.

File: scripts/building/check_enforcer.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+# Define variables
+github_repository = "microsoft/promptflow"
+snippet_debug = 1  # Write debug info to console.
+merge_commit = ""
+loop_times = 30
+github_workspace = os.path.expanduser("~/promptflow/")
+
+# Special cases for pipelines that need to be triggered more or less than default value 1.
+# If 0, the pipeline will not be ignored in check enforcer.
+special_care = {"sdk_cli_tests": 4, "sdk_cli_azure_test": 4}
```
Comment (zhengfeiwang) [inline scripts/building/check_enforcer.py:19]: this might be useful in some scenarios, maybe it will be better to add an example workflow as demonstration and comment it, so that everyone can know what & how to do this

Comment (zhengfeiwang) [reply scripts/building/check_enforcer.py:19]: also, I think it's better to point out what's the key... I just get confused on what should I pick...

**Review by zhengfeiwang — APPROVED** [review]

File: .github/CODEOWNERS [hunk 5]
```diff
@@ -21,3 +21,8 @@
 /scripts/installer/ @microsoft/promptflow-sdk
 /scripts/json_schema/ @microsoft/promptflow-sdk
 /scripts/readme/ @microsoft/promptflow-sdk
+
+/docs/how-to-guides/develop-a-tool/ @microsoft/promptflow-tools
```
Comment (wangchao1230) [inline .github/CODEOWNERS:25]: why do we have these in this PR?

Comment (crazygao) [reply .github/CODEOWNERS]: reverting, merge errors.

File: scripts/building/check_enforcer.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+snippet_debug = 1  # Write debug info to console.
+merge_commit = ""
+loop_times = 30
+github_workspace = os.path.expanduser("~/promptflow/")
+
+# Special cases for pipelines that need to be triggered more or less than default value 1.
+# If 0, the pipeline will not be ignored in check enforcer.
+# Please notice that the key should be the Job Name in the pipeline.
+special_care = {
+    "sdk_cli_tests": 4,
```
Comment (wangchao1230) [inline scripts/building/check_enforcer.py:21]: what's the integer here means? 4 and 0?

**Review by zhengfeiwang — APPROVED** [review]

File: scripts/building/check_enforcer.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
+# Import necessary libraries
```
Comment (wangchao1230) [inline scripts/building/check_enforcer.py:1]: add some docstring on what this tool does

**Review by wangchao1230 — APPROVED** [review]

File: scripts/building/check_enforcer.py (hunk 6 above)
Comment (crazygao) [reply scripts/building/check_enforcer.py:21]: sdk cli tests has 4 matrix triggers. python 3.8, 3.9. 3.10, 3.11.

This special_care number is you cannot trigger less than this number. 

I want these 4 runs all triggered. Then this is 4.
You can also set 3, then if over 3 sdk cli test is triggered, this check is passed.

if you set this value to 0, then no matter how much you are triggered, this check enforcer will just don't check this pipeline success or not.

File: scripts/building/check_enforcer.py (hunk 7 above)
Comment (crazygao) [reply scripts/building/check_enforcer.py:1]: Okay

File: scripts/building/check_enforcer.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+# 2. examples checks: this script calculate the path filters and decide what should be triggered.
+
+# Trigger checks and return the status of the checks:
+# 1. If examples are not correctly generated, fail.
+# 2. If required pipelines are not triggered within 6 rounds of loops, fail.
+#     2.1 (special_care global variable could help on some pipelines that need to bypass the check)
+
+# Check pipelines succeed or not:
+# 1. These pipelines should return status within loop-times rounds.
+# 1. If there is failed pipeline in the triggered pipelines, fail.
```
Comment (zhengfeiwang) [inline scripts/building/check_enforcer.py:16]:
```suggestion
# 2. If there is failed pipeline in the triggered pipelines, fail.
```

Comment (crazygao) [reply scripts/building/check_enforcer.py]: Thanks.

**Review by zhengfeiwang — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]
